qi-lib/flow/core/compiler/deforest/syntax.rkt

```diff
@@ -0,0 +1,298 @@
+             (#%host-expression op)
+             (#%host-expression init)
+             __)))
+  (pattern (#%fine-template
+            ((#%host-expression qi:foldr)
+             (#%host-expression op)
+             (#%host-expression init)
+             _))))
+
+(define-syntax-class fsc-foldl
+  #:attributes (op init)
+  #:literal-sets (fs-literals)
+  #:literals (qi:foldl)
+  (pattern (#%blanket-template
+            ((#%host-expression qi:foldl)
+             (#%host-expression op)
+             (#%host-expression init)
+             __)))
+  (pattern (#%fine-template
+            ((#%host-expression qi:foldl)
+             (#%host-expression op)
+             (#%host-expression init)
+             _))))
+
+(define-syntax-class cad*r-datum
+  #:attributes (countdown)
+  (pattern (~literal qi:car) #:attr countdown #'0)
+  (pattern (~literal qi:cadr) #:attr countdown #'1)
+  (pattern (~literal qi:caddr) #:attr countdown #'2)
+  (pattern (~literal qi:cadddr) #:attr countdown #'3))
+
+(define-syntax-class fsc-list-ref
+  #:attributes (pos name)
+  #:literal-sets (fs-literals)
+  #:literals (qi:list-ref)
+  (pattern (~or (#%fine-template
+                 ((#%host-expression qi:list-ref) _ idx))
+                (#%blanket-template
+                 ((#%host-expression qi:list-ref) __ idx)))
+           #:attr pos #'idx
+           #:attr name #'list-ref)
+  (pattern (~or (esc (#%host-expression cad*r:cad*r-datum))
+                (#%fine-template
+                 ((#%host-expression cad*r:cad*r-datum) _))
+                (#%blanket-template
+                 ((#%host-expression cad*r:cad*r-datum) __)))
+           #:attr pos #'cad*r.countdown
+           #:attr name #'cad*r))
+
+(define-syntax-class fsc-length
+  #:literal-sets (fs-literals)
+  #:literals (qi:length)
+  (pattern (esc
+            (#%host-expression qi:length)))
+  (pattern (#%fine-template
+            ((#%host-expression qi:length) _)))
+  (pattern (#%blanket-template
+            ((#%host-expression qi:length) __))))
+
+(define-syntax-class fsc-empty?
+  #:literal-sets (fs-literals)
+  #:literals (qi:null? qi:empty?)
+  (pattern (esc
+            (#%host-expression (~or qi:empty?
+                                    qi:null?))))
+  (pattern (#%fine-template
+            ((#%host-expression (~or qi:empty?
+                                     qi:null?)) _)))
+  (pattern (#%blanket-template
+            ((#%host-expression (~or qi:empty?
+                                     qi:null?)) __))))
+
+(define-syntax-class fsc-default
+  #:datum-literals (cstream->list)
+  (pattern cstream->list))
+
+(define-syntax-class fsc-syntax
+  (pattern (~or _:fsc-foldr
+                _:fsc-foldl
+                _:fsc-list-ref
+                _:fsc-length
+                _:fsc-empty?
+                _:fsc-default
+                )))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; The actual fusion generator implementation
```

Comment:
I'm not sure this belongs in this module. Everything before this point I see as the responsibility of the expander (even if not currently implemented that way in fact), but this code here is the responsibility of the compiler. How about a new module, say `deforest/fusion.rkt`?

Reply:
Will do.